cip.go

```diff
@@ -129,47 +153,83 @@ func main() {
-	for _, registry := range mfest.Registries {
-		mi[registry.Name] = nil
-	}
-	sc := reg.MakeSyncContext(
-		*manifestPtr,
-		mi,
-		*verbosityPtr,
-		*threadsPtr,
-		*deleteExtraTags,
-		*dryRunPtr,
-		!noSvcAcc,
-		mfest.Registries)
+	// FIXME: Populate access tokens for all registries listed in the manifest.
+	// The data is map[serviceAccountString]token, because the RegistryContext
+	// stores info about which service account to use for the registry when
+	// auth-ing calls against it.
+	//
+	// Then just have the token map in the SyncContext.
 
-	// Read the state of the world; i.e., populate the SyncContext.
-	mkRegistryListingCmd := func(
-		rc reg.RegistryContext) stream.Producer {
+	for _, rc := range sc.RegistryContexts {
+		// at this point, the rc.Name is the root of the repository ("registry"
+		// loosely)
+
+		// (1) GET ACCESS TOKEN via shelling out
 		var sp stream.Subprocess
-		sp.CmdInvocation = reg.GetRegistryListingCmd(
-			rc,
-			sc.UseServiceAccount)
-		return &sp
+		cmd := []string{
+			"gcloud",
+			"auth",
+			"print-access-token",
+		}
+		sp.CmdInvocation = reg.MaybeUseServiceAccount(rc.ServiceAccount, sc.UseServiceAccount, cmd)
+		// check errors
+		sout, _, err := sp.Produce()
+		if err != nil {
+			log.Fatalf("could not get access token for '%s'", rc.ServiceAccount)
+		}
+		token, err := ioutil.ReadAll(sout)
+		if err != nil {
+			log.Fatalf("could not read '%s' for access token", sout)
+		}
+		fmt.Println("KEY:", reg.RootRepo(string(rc.Name)))
+		fmt.Println("VAL:", reg.Token(token))
+		sc.Tokens[reg.RootRepo(string(rc.Name))] = reg.Token(strings.TrimSpace(string(token)))
+		sp.Close()
 	}
 
-	sc.ReadImageNames(mkRegistryListingCmd)
-	mkRegistryListTagsCmd := func(
-		rc reg.RegistryContext, imgName reg.ImageName) stream.Producer {
-		var sp stream.Subprocess
-		sp.CmdInvocation = reg.GetRegistryListTagsCmd(
-			rc,
-			sc.UseServiceAccount,
-			string(imgName))
-		return &sp
+	fmt.Println("hello")
+	fmt.Println(sc.Tokens)
+
+	mkReadRepositoryCmd := func(
+		rc reg.RegistryContext) stream.Producer {
+		var sh stream.HTTP
+
+		tokenKey, domain, repoPath := reg.GetTokenKeyDomainRepoPath(rc.Name)
+
+		httpReq, err := http.NewRequest(
+			"GET",
+			fmt.Sprintf("https://%s/v2/%s/tags/list", domain, repoPath),
+			nil)
+
+		if err != nil {
+			fmt.Println("UNKNOWN ERROR")
+			// handle err... Maybe just print a warning message. Because this
+			// and other operations will take time, and the service is mutating
+			// live (even though this process should be the only one, others might push images that are live.)...
+
+			// TODO: If we detect that we cannot read a repo correctly, we may have to REMOVE that repo's contents from the Manifest, because, e.g., it may be that the manifest defines tags in...
+			//
+			// Basically the idea is that we have to perform safety
+			// modifications to our set data structure info before doing things
+			// like garbage collection, _if we have an incomplete view of the world_.
+		}
```

Comment:
If `err != nil` the function won't stop, are you sure that `httpReq` will be in expected state?

Reply:
I have to handle this error --- I think just aborting entirely is OK. We could get fancy and perform a limited promotion (basically reducing the set data structure to a subset), but I don't think that's what we want. We don't want to print SUCCESS at the end when a http error prevented us from promoting the 1 image we added to our manifest.